### PR TITLE
Allow to skip test code during publishing

### LINKED ISSUE #4365: Beta regressions "failed to parse manifest"
I've filed a few bugs on rust-lang/rust to track all beta regressions there, but these regressions are in fact caused by cargo:

* https://github.com/rust-lang/rust/issues/43660
* https://github.com/rust-lang/rust/issues/43668
* https://github.com/rust-lang/rust/issues/43675
* https://github.com/rust-lang/rust/issues/43685

They are all of the form:

```
error: failed to parse manifest at `/source/Cargo.toml`

Caused by:
  can't find `tests` test, specify test.path
```

Or similar.

### PR DESCRIPTION
**Pull request by matklad**
closes #4365 

I don't have time to review this code carefully right now, but looks like it does the job.

### CODE REVIEW

File: src/cargo/util/toml/targets.rs [hunk 1]
```diff
@@ -38,19 +39,19 @@ pub fn targets(manifest: &TomlManifest,
     }
 
     targets.extend(
-        clean_bins(manifest.bin.as_ref(), package_root, package_name, warnings, has_lib)?
+        clean_bins(manifest.bin.as_ref(), package_root, package_name, warnings, errors, has_lib)?
```
Comment (matklad) [inline src/cargo/util/toml/targets.rs]: Hm, now I am not quite sure that we should apply this logic to binaries as well...

File: tests/test.rs [hunk 2]
```diff
@@ -2732,3 +2732,42 @@ fn cyclic_dev() {
...
+
+            [dependencies]
+            testless = "0.1.0"
+        "#)
+        .file("src/lib.rs", "");
+    p.build();
+
+    assert_that(p.cargo("test"), execs().with_status(0));
+    assert_that(p.cargo("test").arg("--package").arg("testless"),
+                execs().with_status(0));
```
Comment (matklad) [inline tests/test.rs]: I think this last check is something that we actually gain here. 

My undestanding is that with the current cargo it is impossible to do `cargo test --package dependency`, if the dep skipped some tests during publishing. Now it is possible.

Comment (alexcrichton) [inline tests/test.rs]: Ah yeah I think that's ok. This won't work in general due to dev-dependencies anyway (e.g. you can still test the library and it won't link right)

File: src/cargo/util/toml/targets.rs (hunk 1 above)
Comment (alexcrichton) [inline src/cargo/util/toml/targets.rs]: Yeah maybe we can stick to just tests/benchmarks for now?

**Comment by alexcrichton** [comment]
@bors: r+

Oops sorry missed the update here!

**Pull request merged**